File: api/app/selections/preparation.py
import asyncio
import json
from collections.abc import Callable
from datetime import UTC, date, datetime
from enum import StrEnum
from typing import Any, Protocol

from pydantic import Field
from sqlalchemy import Engine, text

from app.domain.errors import (
    CandidateSelectionPreconditionError,
    PersistenceUnavailableError,
    ProjectNotFoundError,
)
from app.domain.models import StrictModel, VentureState
from app.mcp.client import McpCallOutcome, McpClientError
from app.workflows.models import HeadFence


class ProcedureType(StrEnum):
    BUSINESS_REGISTRATION = "BUSINESS_REGISTRATION"
    FOOD_SERVICE_REPORT = "FOOD_SERVICE_REPORT"
    FACILITY_REQUIREMENTS = "FACILITY_REQUIREMENTS"
    HYGIENE_EDUCATION = "HYGIENE_EDUCATION"
    SIGNAGE = "SIGNAGE"
    FIRE_SAFETY = "FIRE_SAFETY"


class PreparationGuideStatus(StrEnum):
    COMPLETE = "COMPLETE"
    REVIEW_REQUIRED = "REVIEW_REQUIRED"
    UNAVAILABLE = "UNAVAILABLE"


class ProcedureRetrievalStatus(StrEnum):
    OK = "OK"
    PARTIAL = "PARTIAL"
    STALE = "STALE"
    NOT_FOUND = "NOT_FOUND"
    ERROR = "ERROR"


class OfficialSourceTrace(StrictModel):
    source_id: str = Field(min_length=1, max_length=128)
    source_ref: str = Field(min_length=1, max_length=2000)
    data_date: date | None
    retrieved_at: datetime
    content_digest: str = Field(pattern=r"^sha256:[0-9a-f]{64}$")


class ProcedureStep(StrictModel):
    procedure_type: ProcedureType
    step_order: int = Field(ge=1)
    title: str = Field(min_length=1)
    required: bool
    authority: str = Field(min_length=1)
    source_date: date
    evidence_id: str = Field(min_length=1)


class ProcedureCoverage(StrictModel):
    procedure_type: ProcedureType
    status: ProcedureRetrievalStatus
    request_id: str | None = None
    steps: list[ProcedureStep]
    missing_fields: list[str]
    conflicts: list[str]
    error_codes: list[str]
    source_trace: list[OfficialSourceTrace]
    evidence_records: list[dict[str, Any]]


class PreparationGuide(StrictModel):
    project_id: str
    selection_id: str
    candidate_id: str
    candidate_type: str
    jurisdiction_code: str
    jurisdiction_display_name: str | None
    as_of: date
    status: PreparationGuideStatus
    procedures: list[ProcedureCoverage]
    source_trace: list[OfficialSourceTrace]
    evidence_records: list[dict[str, Any]]
    human_actions_only: bool = True
    external_submission_performed: bool = False
    generated_at: datetime


class ProcedureMcpClient(Protocol):
    async def call_tool(
        self,
        *,
        venture_project_id: str,
        workflow_run_id: str,
        head: HeadFence,
        tool_name: str,
        arguments: dict[str, Any],
        traceparent: str | None = None,
        timeout_seconds: float = 30.0,
    ) -> McpCallOutcome: ...


class PreparationGuideService:
    def __init__(
        self,
        engine: Engine,
        mcp_client: ProcedureMcpClient,
        *,
        now: Callable[[], datetime] | None = None,
        max_concurrency: int = 4,
    ) -> None:
        if max_concurrency < 1:
            raise ValueError("Preparation guide concurrency must be positive")
        self._engine = engine
        self._mcp_client = mcp_client
        self._now = now or (lambda: datetime.now(UTC))
        self._max_concurrency = max_concurrency

    async def get(
        self,
        *,
        project_id: str,
        selection_id: str,
        user_id: str,
    ) -> PreparationGuide:
        selected = self._load_selection(
            project_id=project_id,
            selection_id=selection_id,
            user_id=user_id,
        )
        state = VentureState.model_validate(selected["state_json"])
        if state.active_case_id != selected["candidate_id"]:
            raise CandidateSelectionPreconditionError(
                "Preparation guide requires the current selected candidate"
            )
        jurisdiction_code = state.area.administrative_code
        if state.area.resolution_status != "RESOLVED" or not jurisdiction_code:
            raise CandidateSelectionPreconditionError(
                "Preparation guide requires a resolved administrative area"
            )
        head = HeadFence(
            workflow_generation=selected["workflow_generation"],
            state_version=selected["state_version"],
            founder_snapshot_id=selected["founder_snapshot_id"],
            area_snapshot_id=selected["area_snapshot_id"],
            evidence_snapshot_id=selected["evidence_snapshot_id"],
            policy_snapshot_id=selected["policy_snapshot_id"],
            index_generation_id=selected["index_generation_id"],
            seed_registry_id=selected["seed_registry_id"],
        )
        generated_at = self._now()
        as_of = generated_at.date()
        semaphore = asyncio.Semaphore(self._max_concurrency)

        async def retrieve(procedure_type: ProcedureType) -> ProcedureCoverage:
            try:
                async with semaphore:
                    outcome = await self._mcp_client.call_tool(
                        venture_project_id=project_id,
                        workflow_run_id=f"preparation-{selection_id}",
                        head=head,
                        tool_name="get_official_procedure",
                        arguments={
                            "jurisdiction_code": jurisdiction_code,
                            "procedure_type": procedure_type.value,
                            "as_of": as_of.isoformat(),
                        },
                        timeout_seconds=30.0,
                    )
            except McpClientError as error:
                return ProcedureCoverage(
                    procedure_type=procedure_type,
                    status=ProcedureRetrievalStatus.ERROR,
                    steps=[],
                    missing_fields=[],
                    conflicts=[],
                    error_codes=[error.mcp_code],
                    source_trace=[],
                    evidence_records=[],
                )
            content = outcome.structured_content
            evidence_records = list(content.get("evidence_records", []))
            evidence_ids = {
                value.get("evidence_id")
                for value in evidence_records
                if isinstance(value, dict)
            }
            unsupported_step_count = 0
            steps: list[ProcedureStep] = []
            for value in content.get("data", []):
                if not isinstance(value, dict):
                    continue
                if value.get("evidence_id") not in evidence_ids:
                    unsupported_step_count += 1
                    continue
                steps.append(ProcedureStep(procedure_type=procedure_type, **value))
            steps.sort(key=lambda value: (value.step_order, value.title))
            error_codes = list(content.get("error_codes", []))
            status = ProcedureRetrievalStatus(outcome.status)
            if unsupported_step_count:
                error_codes.append("PROCEDURE_EVIDENCE_MISSING")
                status = ProcedureRetrievalStatus.PARTIAL
            return ProcedureCoverage(
                procedure_type=procedure_type,
                status=status,
                request_id=outcome.request_id,
                steps=steps,
                missing_fields=list(content.get("missing_fields", [])),
                conflicts=list(content.get("conflicts", [])),
                error_codes=sorted(set(error_codes)),
                source_trace=[
                    OfficialSourceTrace.model_validate(value)
                    for value in content.get("source_trace", [])
                ],
                evidence_records=evidence_records,
            )

        outcomes = await asyncio.gather(*(retrieve(value) for value in ProcedureType))
        source_trace = self._collect_source_trace(outcomes)
        candidate = selected["candidate_json"]
        return PreparationGuide(
            project_id=project_id,
            selection_id=selection_id,
            candidate_id=selected["candidate_id"],
            candidate_type=str(candidate.get("case_type", "UNKNOWN")),
            jurisdiction_code=jurisdiction_code,
            jurisdiction_display_name=state.area.display_name,
            as_of=as_of,
            status=self._guide_status(outcomes),
            procedures=outcomes,
            source_trace=source_trace,
            evidence_records=self._collect_evidence_records(outcomes),
            generated_at=generated_at,
        )

    def _load_selection(
        self,
        *,
        project_id: str,
        selection_id: str,
        user_id: str,
    ) -> dict[str, Any]:
        with self._engine.connect() as connection:
            row = connection.execute(
                text(
                    """
                    SELECT cs.selection_id, cs.candidate_id, cs.candidate_json,
                           s.state_json, h.workflow_generation, h.state_version,
                           h.founder_snapshot_id, h.area_snapshot_id,
                           h.evidence_snapshot_id, h.policy_snapshot_id,
                           h.index_generation_id, h.seed_registry_id
                    FROM candidate_selections cs
                    JOIN venture_projects p ON p.project_id=cs.project_id
                    JOIN venture_states s
                      ON s.project_id=p.project_id
                     AND s.state_version=p.current_state_version
                    JOIN project_heads h ON h.project_id=p.project_id
                    WHERE cs.project_id=:project_id
                      AND cs.selection_id=:selection_id
                      AND cs.owner_user_id=:user_id
                      AND p.owner_user_id=:user_id
                    """
                ),
                {
                    "project_id": project_id,
                    "selection_id": selection_id,
                    "user_id": user_id,
                },
            ).mappings().one_or_none()
        if row is None:
            raise ProjectNotFoundError("Candidate selection was not found")
        value = dict(row)
        for field in ("candidate_json", "state_json"):
            if isinstance(value[field], str):
                value[field] = json.loads(value[field])
        return value

    @staticmethod
    def _guide_status(outcomes: list[ProcedureCoverage]) -> PreparationGuideStatus:
        if all(value.status == "OK" and value.steps for value in outcomes):
            return PreparationGuideStatus.COMPLETE
        if all(value.status in {"ERROR", "NOT_FOUND"} and not value.steps for value in outcomes):
            return PreparationGuideStatus.UNAVAILABLE
        return PreparationGuideStatus.REVIEW_REQUIRED

    @staticmethod
    def _collect_source_trace(
        outcomes: list[ProcedureCoverage],
    ) -> list[OfficialSourceTrace]:
        unique: dict[tuple[str, str, str], OfficialSourceTrace] = {}
        for outcome in outcomes:
            for source in outcome.source_trace:
                key = (
                    source.source_id,
                    source.source_ref,
                    source.content_digest,
                )
                unique[key] = source
        return [unique[key] for key in sorted(unique)]

    @staticmethod
    def _collect_evidence_records(
        outcomes: list[ProcedureCoverage],
    ) -> list[dict[str, Any]]:
        unique: dict[str, dict[str, Any]] = {}
        for outcome in outcomes:
            for evidence in outcome.evidence_records:
                evidence_id = evidence.get("evidence_id")
                if isinstance(evidence_id, str) and evidence_id:
                    unique[evidence_id] = evidence
        return [unique[evidence_id] for evidence_id in sorted(unique)]


class UnavailablePreparationGuideService:
    async def get(self, **_: Any) -> PreparationGuide:
        raise PersistenceUnavailableError("Preparation guide dependencies are unavailable")

File: src/uiSimulation/preparation.ts
import type { PreparationGuide, PreparationProcedure, ProcedureType, Project, ResultCandidate } from '../apiClient'

const asOf = '2026-08-25'
const retrievedAt = '2026-08-25T07:00:00Z'

const officialSources: Record<ProcedureType, { id: string; ref: string }> = {
  FACILITY_REQUIREMENTS: { id: 'source:food-facility', ref: 'https://www.foodsafetykorea.go.kr' },
  HYGIENE_EDUCATION: { id: 'source:food-hygiene', ref: 'https://www.foodsafetykorea.go.kr' },
  FOOD_SERVICE_REPORT: { id: 'source:gov-food-report', ref: 'https://www.gov.kr' },
  BUSINESS_REGISTRATION: { id: 'source:hometax-registration', ref: 'https://www.hometax.go.kr' },
  SIGNAGE: { id: 'source:seongdong-signage', ref: 'https://www.sd.go.kr' },
  FIRE_SAFETY: { id: 'source:nfa-fire', ref: 'https://www.nfa.go.kr' },
}

function procedure(
  type: ProcedureType,
  steps: Array<{ title: string; required: boolean; authority: string }>,
  options: { status?: PreparationProcedure['status']; missingFields?: string[] } = {},
): PreparationProcedure {
  const source = officialSources[type]
  return {
    procedure_type: type,
    status: options.status ?? 'OK',
    steps: steps.map((step, index) => ({
      procedure_type: type,
      step_order: index + 1,
      title: step.title,
      required: step.required,
      authority: step.authority,
      source_date: asOf,
      evidence_id: `evidence:procedure:${type.toLowerCase()}:${index + 1}`,
    })),
    missing_fields: options.missingFields ?? [],
    conflicts: [],
    error_codes: [],
    source_trace: [{
      source_id: source.id,
      source_ref: source.ref,
      data_date: asOf,
      retrieved_at: retrievedAt,
      content_digest: `sha256:${type.charCodeAt(0).toString(16).padStart(2, '0').repeat(32).slice(0, 64)}`,
    }],
    evidence_records: [],
  }
}

const procedures: PreparationProcedure[] = [
  procedure('FACILITY_REQUIREMENTS', [
    { title: '조리장·급수·환기·세척·화장실 시설기준을 점포 도면과 현장에서 대조', required: true, authority: '성동구청 위생 담당 부서' },
    { title: '배수·환기·냉장보관·손씻기 설비의 현재 상태와 공사 필요 범위를 확인', required: true, authority: '성동구청 위생 담당 부서' },
    { title: '인테리어 공사 전에 변경 계획을 들고 시설기준 적용사항을 사전 문의', required: false, authority: '성동구청 위생 담당 부서' },
  ]),
  procedure('HYGIENE_EDUCATION', [
    { title: '신규 영업자 위생교육 대상과 인정 교육기관을 확인', required: true, authority: '식품위생교육기관' },
    { title: '신규 영업자 위생교육 이수', required: true, authority: '식품위생교육기관' },
    { title: '영업신고에 사용할 교육 수료증을 준비', required: true, authority: '식품위생교육기관' },
  ]),
  procedure('FOOD_SERVICE_REPORT', [
    { title: '휴게음식점 영업신고 준비', required: true, authority: '성동구청 위생 담당 부서' },
    { title: '구비서류와 관할 접수처를 최종 확인', required: true, authority: '성동구청 위생 담당 부서' },
    { title: '시설기준 확인과 위생교육 이수 후 영업신고 서류를 제출', required: true, authority: '성동구청 위생 담당 부서' },
  ]),
  procedure('BUSINESS_REGISTRATION', [
    { title: '사업자등록에 필요한 임대차계약서·영업신고 관련 서류를 준비', required: true, authority: '관할 세무서 또는 국세청 홈택스' },
    { title: '사업자등록 신청 시 업태·종목과 사업장 주소를 확인해 제출', required: true, authority: '관할 세무서 또는 국세청 홈택스' },
    { title: '사업자등록증 발급 뒤 카드가맹·계좌·세금계산서 설정에 사용할 정보를 확인', required: false, authority: '국세청 홈택스' },
  ]),
  procedure('SIGNAGE', [
    { title: '설치할 간판의 종류·크기·조명·위치를 건물 외관 계획에 표시', required: true, authority: '성동구청 옥외광고물 담당 부서' },
    { title: '신고·허가 대상 여부와 수량 제한을 설치 전에 확인', required: true, authority: '성동구청 옥외광고물 담당 부서' },
    { title: '건물 관리규정과 임대인의 간판 설치 동의 범위를 함께 확인', required: false, authority: '건물 관리주체 및 임대인' },
  ]),
  procedure('FIRE_SAFETY', [
    { title: '점포 면적·층·수용인원·건물 용도에 따른 소방 적용사항을 확인', required: true, authority: '성동소방서' },
    { title: '소화기·유도등·비상구·감지기 등 현재 소방시설 상태를 현장에서 확인', required: true, authority: '성동소방서 또는 소방 점검 주체' },
    { title: '인테리어 공사로 소방시설 위치나 구획이 바뀌면 공사 전에 협의', required: true, authority: '성동소방서 또는 소방시설 관계자' },
  ]),
]

export function buildSeongsuPreparationGuide(
  project: Project,
  selectionId: string,
  candidate: ResultCandidate | undefined,
): PreparationGuide {
  return {
    project_id: project.project_id,
    selection_id: selectionId,
    candidate_id: candidate?.candidate_id ?? 'candidate:unselected',
    candidate_type: candidate?.case_type ?? 'INDEPENDENT',
    jurisdiction_code: project.state?.area.area_id ?? 'area:unresolved',
    jurisdiction_display_name: project.state?.area.display_name ?? '서울특별시 성동구 성수동',
    as_of: asOf,
    status: 'COMPLETE',
    procedures,
    source_trace: Object.values(officialSources).map((source, index) => ({
      source_id: source.id,
      source_ref: source.ref,
      data_date: asOf,
      retrieved_at: retrievedAt,
      content_digest: `sha256:${String(index + 1).repeat(64).slice(0, 64)}`,
    })),
    evidence_records: [],
    human_actions_only: true,
    external_submission_performed: false,
    generated_at: retrievedAt,
  }
}